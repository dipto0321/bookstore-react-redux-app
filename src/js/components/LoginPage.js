import React, { useState } from 'react';
import { connect } from 'react-redux';
import { login } from '../thunks/user';
import InputWrapper from './InputWrapper';
import Header from './Header';

export const LoginPage = ({ login, error, history }) => {
  const [emailOrUsername, setEmailOrUsername] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState(error);

  const loginUser = (e) => {
    e.preventDefault();
    setEmailOrUsername('');
    setPassword('');
    login({ email_or_username: emailOrUsername, password })
      .then(() => {
        history.push('/');
      })
      .catch(() => setErrorMessage(error));
  };

  return (
    <div>
      <Header />
      <div className="form-wrapper">
        <form className="form user-form">
          {error && <div className="error">{errorMessage}</div>}
          <InputWrapper
            inputValue={emailOrUsername}
            labelValue="Username/Email:"
            setInput={setEmailOrUsername}
            type="text"
            inputId="email"
          />
          <InputWrapper
            inputValue={password}
            labelValue="Password:"
            setInput={setPassword}
            type="password"
            inputId="password"
          />
          <div className="btn-wrapper">
            <button
              type="submit"
              className="user-form__btn"
              onClick={loginUser}
            >
              Login
            </button>
            <button
              type="button"
              className="user-form__btn"
              onClick={() => history.push('/signup')}
            >
              Sign Up
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  error: state.error,
});

export default connect(
  mapStateToProps,
  { login },
)(LoginPage);
