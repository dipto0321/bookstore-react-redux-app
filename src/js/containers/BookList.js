import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { switchAddBookMode, removeBook, switchProgressUpdate } from '../actions/books';
import { changeFilter } from '../actions/filter';
import Book from '../components/Book';
import CategoryFilter from '../components/CategoryFilter';
import ProgressUpdateModal from '../components/ProgressUpdateModal';

export const BookList = (
  {
    books,
    changeFilter,
    filter,
    removeBook,
    switchAddBookMode,
    progressUpdateMode,
    switchProgressUpdate,
  },
) => (
  <div>
    <ProgressUpdateModal
      progressUpdateMode={progressUpdateMode}
      switchProgressUpdate={switchProgressUpdate}
    />
    <div className="booklist-controls">
      <CategoryFilter
        filter={filter}
        handleChange={optionValue => changeFilter(optionValue)}
      />
      <button
        className="add-book-btn btn-sm"
        type="button"
        onClick={switchAddBookMode}
      >
          Add Book
      </button>
    </div>
    <div>
      {books.map(book => (
        <Book
          key={book.id}
          book={book}
          handleRemove={id => removeBook(id)}
        />
      ))}
    </div>
  </div>
);

const mapStateToProps = state => ({
  books: state.books.filter(book => (state.filter === '' ? true : book.category === state.filter)),
  filter: state.filter,
  progressUpdateMode: state.progressUpdateMode,
});

BookList.propTypes = {
  books: PropTypes.instanceOf(Object).isRequired,
  changeFilter: PropTypes.func.isRequired,
  filter: PropTypes.string.isRequired,
  removeBook: PropTypes.func.isRequired,
  switchAddBookMode: PropTypes.func.isRequired,
  progressUpdateMode: PropTypes.instanceOf(Object).isRequired,
  switchProgressUpdate: PropTypes.func.isRequired,
};

export default connect(
  mapStateToProps,
  {
    removeBook, changeFilter, switchAddBookMode, switchProgressUpdate,
  },
)(BookList);
