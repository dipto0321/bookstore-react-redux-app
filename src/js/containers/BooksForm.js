import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import categories from '../data/bookCategories';

const BooksForm = () => {
  return (
    <form>
      <input type="text"/>
      <select>
        {
          categories.map(category => (
            <option
              key={category}
              value={category}
            >
              { category }
            </option>
          ))
        }
      </select>
      <button 
        type="submit"      
      >
        Add Book
      </button>
    </form>
  )
};

export default BooksForm;
