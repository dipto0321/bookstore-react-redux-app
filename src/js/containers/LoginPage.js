import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { login } from '../thunks/currentUser';

export const LoginPage = ({ login, error, history }) => {
  const [emailOrUsername, setEmailOrUsername] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState(error);

  useEffect(() => {
    setErrorMessage(error);
  }, [error]);

  const loginUser = (e) => {
    e.preventDefault();
    login({ email_or_username: emailOrUsername, password })
      .then(() => {
        setEmailOrUsername('');
        setPassword('');
        setErrorMessage(error);
        history.push('/');
      }).catch(() => {});
  };

  return (
    <form>
      {
        error && <div className="error">{errorMessage}</div>
      }
      <div>
        <label htmlFor="email">Username/Email:</label>
        <input
          type="email"
          id="email"
          value={emailOrUsername}
          onChange={e => setEmailOrUsername(e.target.value)}
        />
      </div>
      <div>
        <label htmlFor="password">Password:</label>
        <input
          type="password"
          id="password"
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
      </div>
      <button
        type="submit"
        onClick={loginUser}
      >
        Login
      </button>
    </form>
  );
};

const mapStateToProps = state => ({
  error: state.error,
});

export default connect(mapStateToProps, { login })(LoginPage);
