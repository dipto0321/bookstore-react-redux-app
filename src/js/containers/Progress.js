import React from 'react';
import CircularProgressbar from 'react-circular-progressbar';

const Progress = () => {
  const percentage = 66;
  return (
    <div className="progress-body">
      <div className="progress-bar">
        <CircularProgressbar percentage={percentage} />
      </div>
      <div className="progress-percentage">
        <p className="percent-complete">{`${percentage}%`}</p>
        <p className="completed">Completed</p>
      </div>
    </div>
  );
};

export default Progress;
