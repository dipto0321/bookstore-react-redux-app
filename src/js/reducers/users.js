import { LIST_USERS } from '../actions/actionTypes';

export default (state = [], action) => {
  if (action.type === LIST_USERS) {
    return action.users;
  }
  return state;
};
