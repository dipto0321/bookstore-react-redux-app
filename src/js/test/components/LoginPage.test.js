import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../containers/LoginPage';
import users from '../fixtures/users';

describe('LoginPage component', () => {
  let wrapper;
  let login;
  let history;

  beforeEach(() => {
    login = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<LoginPage error={null} login={login} history={history} />);
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('successful Login', () => {
    test('should call authenticate function', () => {
      wrapper.find('input[type="email"]').prop('onChange')({
        target: { value: users[0].username },
      });
      wrapper.find('input[type="password"]').prop('onChange')({
        target: { value: users[0].password },
      });
      wrapper.find('button').simulate('click', {
        preventDefault: () => {},
      });
      expect(login).toHaveBeenLastCalledWith({
        email_or_username: users[0].username,
        password: users[0].password,
      });
      expect(history.push).toHaveBeenLastCalledWith('/');
    });
  });

  describe('wrong password', () => {
    test('should show error message', () => {
      wrapper.find('input[type="email"]').simulate('change', { target: { value: users[0].email } });
      wrapper.find('input[type="password"]').simulate('change', { target: { value: 'somePassword' } });
      wrapper.find('button').simulate('click', { preventDefault: () => {} });
      wrapper.setProps({ error: 'Invalid credentials' });

      expect(wrapper.find('div.error').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
