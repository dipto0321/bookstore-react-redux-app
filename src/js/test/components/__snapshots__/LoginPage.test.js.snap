// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage component should render correctly 1`] = `
<form>
  <div>
    <label
      htmlFor="email"
    >
      Username/Email:
    </label>
    <input
      id="email"
      onChange={[Function]}
      type="email"
      value=""
    />
  </div>
  <div>
    <label
      htmlFor="password"
    >
      Password:
    </label>
    <input
      id="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <button
    onClick={[Function]}
    type="submit"
  >
    Login
  </button>
</form>
`;

exports[`LoginPage component wrong password should show error message 1`] = `
<form>
  <div
    className="error"
  />
  <div>
    <label
      htmlFor="email"
    >
      Username/Email:
    </label>
    <input
      id="email"
      onChange={[Function]}
      type="email"
      value="ryto@gmail.com"
    />
  </div>
  <div>
    <label
      htmlFor="password"
    >
      Password:
    </label>
    <input
      id="password"
      onChange={[Function]}
      type="password"
      value="somePassword"
    />
  </div>
  <button
    onClick={[Function]}
    type="submit"
  >
    Login
  </button>
</form>
`;
